lib/server.cpp

```diff
@@ -0,0 +1,286 @@
+	if(streamPolicy.first == "none") {
+		streamPolicy = std::pair<string, int>(name, timeout);
+	}
+	return false; // can only change from default
+}
+
+bool Socialite::Server::setupSSL(string cert, string key) {
+	SSL_load_error_strings();
+	OpenSSL_add_ssl_algorithms();
+	method = (SSL_METHOD*) TLSv1_2_server_method(); // TLSv1_server_method
+	ctx = SSL_CTX_new(method);
+
+	if(!ctx) {
+		return false;
+	}
+
+	char* password = "myaga";
+	SSL_CTX_set_default_passwd_cb_userdata(ctx, (void*) password);
+	SSL_CTX_set_default_passwd_cb(ctx, pem_passwd_cb);
+
+	if(!SSL_CTX_use_PrivateKey_file(ctx, key.c_str(), SSL_FILETYPE_PEM)) {
+		return false;
+	}
+
+	std::ifstream certTest(cert);
+	string line;
+	int certCount = 0;
+	while(std::getline(certTest, line)) {
+		if(line == "-----BEGIN CERTIFICATE-----") {
+			certCount++;
+		}
+	}
+
+	if(certCount > 1) { // we've got a chain
+		if(SSL_CTX_use_certificate_chain_file(ctx, cert.c_str()) != 1) {
+			return false;
+		}
+	}
+	else {
+		if(SSL_CTX_use_certificate_file(ctx, cert.c_str(), SSL_FILETYPE_PEM) != 1) {
+			return false;
+		}
+	}
+
+	if (!SSL_CTX_check_private_key(ctx)) {
+		return false;
+	}
+
+	return true;
+}
+
+bool Socialite::Server::setupServer(int config) {
+	mainSock = socket(AF_INET, SOCK_STREAM, 0);
+	if(mainSock < 0) {
+		printf("ERROR: socket\n");
+		return false;
+	}
+	int reuse_true = 1;
+
+	// no timeout between restarting servers
+	setsockopt(mainSock, SOL_SOCKET, SO_REUSEADDR, &reuse_true, sizeof(reuse_true));
+
+	if(config & S_KEEP_ALIVE) {
+		// use keep-alive to ping the connection every now and again to make sure
+		// it doesn't go dead
+		setsockopt(mainSock, SOL_SOCKET, SO_KEEPALIVE, &reuse_true, sizeof(reuse_true));
+	}
+	// this might be an OS X only thing: SO_REUSEPORT
+	//setsockopt(mainSock, SOL_SOCKET, SO_REUSEPORT, &reuse_true, sizeof(reuse_true));
+
+	struct sockaddr_in addr; // internet socket address data structure
+	addr.sin_family = AF_INET;
+	addr.sin_port = htons(port); // byte order is significant
+	addr.sin_addr.s_addr = INADDR_ANY; // listen to all interfaces
+	//addr.sin_addr = INADDR_ANY; // listen to all interfaces
+
+	if(bind(mainSock, (struct sockaddr*)&addr, sizeof(addr)) < 0) {
+		printf("errno: %i\n", errno);
+		if(errno == 13) {
+			printf("Permission Denined! Couldn't bind to %i. Are you maybe you didn't launch as root.\n", port);
+		}
+		printf("Setup failed\n");
+		return false;
+	}
+
+	if(listen(mainSock, waitingCount) < 0) {
+		printf("ERROR: listen\n");
+		return false;
+	}
+	return true;
+}
+
+int Socialite::Server::pem_passwd_cb(char *buf, int size, int rwflag, void *password) {
+	strncpy(buf, (char*) password, size);
+	buf[size - 1] = '\0';
+	return (strlen(buf));
+}
+
+void Socialite::Server::stop() {
+	if(running) {
+		running = false;
+		// should unbind and then mainloop `server_listening_thread_run`
+		// and will close 
+		std::cout << "Socialite::Server::stop closing mainsock " << mainSock << " on port " << port << "\n";
```

Comment:
Would be nice if this logging was optional. Might not always want this going to stdout